### ISSUE TITLE
Authentication, authorization and session backing store need to be separated

### ISSUE THREAD

**Issue by lukehutch** [eclipse-vertx/vertx-auth#272]
### Version

* vert.x core: 4.0-SNAPSHOT
* vert.x auth: 4.0-SNAPSHOT

### Context

As far as I can tell, there are numerous limitations to the current authentication, authorization and session management classes in Vertx:

* Currently OAuth2 support (I'll use `GoogleAuth` as the specific example) is ephemeral, in the sense that it can be stored in the local session, but there's no way to directly persist these sessions in a backing store such as MongoDB. This means that:
  * If the Vertx server restarts, the user has to authenticate to Google again. This adds a few seconds to the first connection to the server, while the Google authentication cookie is fetched again -- maybe not such a big deal, but not ideal.
  * More importantly, it's not possible to associate persistent state with the Google login, in particular roles and permissions.
* However, `MongoAuth` supports only username/password authentication, not OAuth2.
  * As an aside, in a separate limitation, `MongoAuth` does support roles and permissions, but they can only be set when `insertUser` is called, without accessing the Mongo `users` database directly (as far as I have been able to tell).
* With `GoogleAuth`, there is also no easy way to add a "post-successful login" handler that fetches user account information by calling `((AccessToken) ctx.user()).userInfo(handler)`. This is needed to fetch the user email address, so that role-based access control can use a durable unique identifier (the email address).

The only solution I have been able to find is to chain together two auth providers -- to add `GoogleAuth` and then `MongoAuth` to every route. I have to override the redirect after successful Google login by using `.extraParams(new JsonObject().put("state", postLoginPath))` to direct the response to a separate handler, mounted at `postLoginPath` (however, this prevents correct redirect after successful login). I then call `((AccessToken) ctx.user()).userInfo(handler)` from `postLoginPath` to get the email address, then I have to find or create a user account with `MongoAuth`, and add the roles and permissions to the created user account.  

If more than one OAuth provider is to be used, and multiple OAuth providers are to be linked to a single user account (e.g. by email address), this situation starts to get really messy.

### Proposed solution

* Authentication, authorization, and user session backing storage should be completely separated. For example, `MongoAuth` should know nothing about username/password authentication per se, and `GoogleAuth` should allow for persistent backing of the user login. There should be separate `OAuth2Auth`, `UsernamePasswordAuth` and `MongoUserBackingStore` objects that work with `LocalSessionStore`/`UserSessionHandler`.
* `LocalSessionStore` should act as a write-through cache for `MongoUserBackingStore`, for speed, so that MongoDB is only called out to when absolutely needed (e.g. to query a user for the first time, or in a clustered environment).
* There should be post-authentication hooks that can be called, in particular:
  * to fetch additional metadata, e.g. a user's email address or profile information, from an OAuth server after authentication but before logging the user into the Vertx server, and
  * to handle differently each of the following situations:
    * login of existing user, with an OAuth token or password that exactly matches what is stored in the backing store (standard login);
    * login of existing user, when the OAuth token is valid but has changed (e.g. to trigger the Vertx server to pull down any metadata about the remote user account, so that the local user object can be updated);
    * creation of new user account, e.g. to add new default roles and permissions *only* when a new user account has just been created given a first-time valid OAuth2 login.
    * adding of a new OAuth login of different type to the same user object (e.g. so that you can link the OAuth logins of both Facebook and Google to the same user object in the backing store).

**Comment by pmlopes** [comment]
Hi @lukehutch I agree with much of what you wrote, could you draft some basic interfaces to have the idea how the whole API would work/look? The main concern at the moment is that with 4.0 we can break and add new stuff, but I'd like to have a "easy" upgrade path too...

**Comment by pmlopes** [comment]
```java
interface Authentication {
  void authenticate(JsonObject input, Handler<AsyncResult<User> callback);
}

class Oauth2 implements Authentication {
}
```

How would you plug the storage? authZ to this? how would the events you listed be triggered?

**Comment by lukehutch** [comment]
I would start with

```
interface PersistentSessionStore extends SessionStore {
    // ...
}

interface WriteThroughCache {
    /** Register volatile and persistent session stores. */
    void register(SessionStore volatileSessionStore, PersistentSessionStore persistentSessionStore);

    /**
     * Write item to volatile session store, then asynchronously write through to persistent store.
     * Deletes entry from cache if cacheItem is null.
     */
    void update(String id, JsonObject cacheItem, Handler<Boolean> writeThroughFinishedHandler);

    /** Get item from volatile cache -- fails if id is not present in volatile session store. */
    JsonObject get(String id);

    /**
     * Invalidate volatileSessionStore copy of cacheItem with given id by pulling it back in from
     * persistentSessionStore to volatile store, if it exists in persistent store. (Useful if persistent copy
     * is updated externally.)
     */
    void invalidate(String id, Handler<JsonObject> persistentCopyHandler);
}
```

Then I would create separate authentication and authorization interfaces

```
interface AuthenticationHandler {
    void registerCache(WriteThroughCache cache);

    void registerUserInfoFetcher(JsonObject principal, Handler<JsonObject> userInfoHandler);

    void registerUserIdFetcher(JsonObject userInfo, Handler<String> userIdHandler);

    void authenticate(RoutingContext ctx, Handler<JsonObject> principalHandler);

    /**
     * In a default implementation, check the cache to see if a user with the id from the session exists
     * in the cache, and if so, continue to authenticate that user. Otherwise call authenticate() to get
     * the principal, then call the user info fetcher (if registered) to exchange the principal for a user info
     * object, then call the user id fetcher (if registered) to get the user id. (Otherwise, maybe just use
     * the first field in the principal or the user info object, hoping it is the id??) Then write an entry to
     * the write through cache for the user (with an empty "authorities" field by default).
     */
    void doAuthenticate(RoutingContext ctx, Handler<Boolean> authenticatedHandler);
}

interface AuthorizationHandler {
    void registerCache(WriteThroughCache cache);

    /** Require the listed authority */
    void authorize(RoutingContext ctx, String requiredAuthority, Handler<Boolean> authorizedHandler);

    /** Require all the listed authorities */
    void authorizeRequiringAll(RoutingContext ctx, Collection<String> requiredAuthorities, Handler<Boolean> authorizedHandler);

    /** Require any of the listed authorities */
    void authorizeRequiringAny(RoutingContext ctx, Collection<String> requiredAuthorities, Handler<Boolean> authorizedHandler);

    /** Add an authority to the "authorities" field of the user record. Should also invalidate the cache for the user id. */
    void addAuthority(String userId, String authority);

    /** Remove an authority from the "authorities" field of the user record. Should also invalidate the cache for the user id. */
    void removeAuthority(String userId, String authority);
}
```

The `authorities` field should be a map, e.g. `{ "authorities": { "superuser": true } }`

I probably got some things wrong above, because I'm still learning the async handler model of Vert.x, and I'm still getting my head around how authZ and authN work in Vert.x, but it might provide a start...

The part I didn't include was how to add multiple authentication types. A given system might need to support several OAuth login types (e.g. Google and Facebook), and maybe also user/password authentication. There should be support for linking multiple OAuth types via a common user id field (hence the methods to fetch a user info object given a principal, and then to fetch a user id given a user info object).

**Comment by lukehutch** [comment]
This is what I have so far, as far as an implementation that tries to work with the current framework. It is a monstrosity...

https://pastebin.com/054RQ7ym

It is lacking a few things, e.g. the ability to add or revoke authorities from the database cache (which would also need to call `AbstractUser::revoke`), but it basically works.

I also can't figure out how to make use of the local session store from the `OAuth2RBAC` implementation, in order to cache the authorities/permissions and user info, since `OAuth2RBAC` does not have access to the session that the `User`/`AccessToken` object came from -- it's SAM takes only the parameters `AccessToken user, String requiredAuthority, Handler<AsyncResult<Boolean>> handler`, and none of them seem to give access to the session or the session store. (It seems like the `User` object should provide a `.getSession()` method.) To call the store directly, the `OAuth2RBAC` implementation would have to figure out the session ID, which has to be obtained from a cookie, which has to be obtained from the request (and the provided parameters don't have access to the request either).

Another issue with the `OAuth2RBAC` implementation is that since `user` does not give access to the session ID, I have to use `user.principal().getString("id_token")` as a proxy for the "session id". I map those ids to the email address in a `ConcurrentHashMap`, so that I can cache a mapping from the `id_token` to the email address (the UID for the user). But `id_token` changes with each OAuth2 login, so the mapping in the cache can grow without bound. It it possible to write custom metadata to the `User` object (such as the user's email address) and have that stored in the session? I can't see how to do that.

**Comment by lukehutch** [comment]
For contrast, this is how I had to solve the problem -- I couldn't use the current RBAC framework to add caching, because there is no access to the current session, so I had to build my own caching handlers from scratch.

https://pastebin.com/xxhUCuPb

**Comment by vinscom** [comment]
I think AuthN/AuthZ, Session and User Profile are three independent areas.
As part of Auth flow we are authenticating user identity and in process we may get additional information from authenticating system accessible from User object.
But user object should not be considered as user profile. And if application has need to maintain user profile, then application has to maintain user profile on its own and has to make sure that user profile is updated during user login. 
In session based application, session will be authenticated as part of auth process and session will be given right permissions.
In non-session based application, during login process, application can generate its own JWT with all the required information to access application.
It is never a good idea to directly depend on multiple auth provider. In case there is need to connect to multiple providers then it is better to have an intermediate system on which application can depend and this intermediate system will handle differences between different providers (Like Keycloak or OAuth).
As for application restart is concern, you can use cluster setup where session is saved in cluster. And as user object is serialisable, you can save user object in session and read same user in different server.

**Comment by lukehutch** [comment]
@vinscom sure, but there are still many changes that need to be made to the current system to make it more flexible. For example, `vertx-auth-mongodb` backs its sessions directly in MongoDB, but that backend is not available to the OAuth2 providers if you want to use them to cache the OpenID Connect metadata or permissions/roles, for example. 

My main point is that the whole system needs to be modularized and made more generic and pluggable, so that AuthN, AuthZ, session persistence, credential caching, etc. can all be mixed and matched more flexibly. It's halfway there right now, but even standard usecases are not workable with the current classes. I'm having to rewrite my own handlers from scratch to do a lot of things that are really standard for identity and AuthZ/AuthN management.

I don't want to pay the price of adding Keycloak etc. -- too heavyweight. There needs to be a better lightweight set of building blocks in Vertx.

**Comment by pmlopes** [comment]
Fixed by #309
